src/main.cpp

```diff
@@ -5713,6 +5713,9 @@ bool SendMessages(CNode* pto)
             pto->vAddrToSend.clear();
             if (!vAddr.empty())
                 pto->PushMessage(NetMsgType::ADDR, vAddr);
+            // we only send the big addr message once
+            if (pto->vAddrToSend.capacity() > 40)
+                pto->vAddrToSend = {};
```

Comment:
Wow, is the temporary-swap still necessary? I was so happy thinking we left that behind with C++98. I think a move assignment would work -- for compatible allocators its time complexity is documented as constant in the size of the other vector, which is only possible if it adopts the other vector's storage.

I'm inclined to go with the straightforward `shrink_to_fit`. It's non-binding specifically to allow for optimizations, and no sane implementation would refuse to deallocate 40KiB as an "optimization"...